crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs

```diff
@@ -0,0 +1,900 @@
+    OxcDiagnostic::warn("Switch to `Set`")
+        .with_help("should be a `Set`, and use .has()` to check existence or non-existence.")
+        .with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct PreferSetHas;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Prefer `Set#has()` over `Array#includes()` when checking for existence or non-existence.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Set#has() is faster than Array#includes().
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// const array = [1, 2, 3];
+    /// const hasValue = value => array.includes(value);
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// const set = new Set([1, 2, 3]);
+    /// const hasValue = value => set.has(value);
+    /// ```
+    /// ```js
+    /// const array = [1, 2, 3];
+    /// const hasOne = array.includes(1);
+    /// ```
+    PreferSetHas,
+    perf,
+    dangerous_fix
+);
+
+fn is_includes_call(member_expr: &MemberExpression) -> bool {
+    if member_expr.is_computed() || member_expr.optional() {
+        return false;
+    }
+
+    let Some(callee_name) = member_expr.static_property_name() else {
+        return false;
+    };
+
+    callee_name == "includes"
+}
+
+fn is_array_of_or_from(name: &str, callee: &MemberExpression) -> bool {
+    if name != "from" && name != "of" {
+        return false;
+    }
+
+    return callee.object().get_identifier_reference().map_or(false, |ident| ident.name == "Array");
+}
+
+fn is_kind_of_array_expr(expr: &Expression) -> bool {
+    match expr {
+        Expression::NewExpression(new_expr) => {
+            new_expr.callee.get_identifier_reference().map_or(false, |ident| ident.name == "Array")
+        }
+        Expression::CallExpression(call_expr) => {
+            let Some(callee) = call_expr.callee.get_member_expr() else {
+                return call_expr.callee_name().map_or(false, |name| name == "Array");
+            };
+
+            if callee.is_computed() || callee.optional() {
+                return false;
+            }
+
+            let Some(name) = callee.static_property_name() else { return false };
+
+            is_array_of_or_from(name, callee) || ARRAY_METHODS_RETURNS_ARRAY.contains(name)
+        }
+        Expression::ArrayExpression(_) => true,
+        _ => false,
+    }
+}
+
+fn is_multiple_calls(node: &AstNode, ctx: &LintContext, root_scope_id: ScopeId) -> bool {
+    let mut id = node.id();
+    let mut was_in_root_scope = node.scope_id() == root_scope_id;
+    loop {
+        let Some(parent) = ctx.nodes().parent_node(id) else { return false };
+        let parent_scope = parent.scope_id();
+        if was_in_root_scope && parent_scope != root_scope_id {
+            return false;
+        }
+        was_in_root_scope = parent_scope == root_scope_id;
+        let parent_kind = parent.kind();
+        if matches!(
```

Comment:
would it break anything to move this check up earlier? checking `AstKind` is faster than getting the parent node, so doing fewer calls to `parent_node` should be faster.

Reply:
Here i'm looping over all the parent and only existing the loop if we exist the scope of the variable declaration or if i found a match for multiple call. So i'm not sure if i can do any change here